src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.RunAsClient.Windows.cs

```diff
@@ -13,23 +13,36 @@ namespace System.IO.Pipes.Tests
 {
     public partial class NamedPipeTest_RunAsClient : RemoteExecutorTestBase
     {
-        [Fact]
+        [Theory]
+        [InlineData(TokenImpersonationLevel.None)]
+        [InlineData(TokenImpersonationLevel.Anonymous)]
+        [InlineData(TokenImpersonationLevel.Identification)]
+        [InlineData(TokenImpersonationLevel.Impersonation)]
+        [InlineData(TokenImpersonationLevel.Delegation)]
         [PlatformSpecific(TestPlatforms.Windows)]  // Uses P/Invokes
         [ActiveIssue(22271, TargetFrameworkMonikers.UapNotUapAot)]
-        public async Task RunAsClient_Windows()
+        public async Task RunAsClient_Windows(TokenImpersonationLevel tokenImpersonationLevel)
         {
             string pipeName = GetUniquePipeName();
             using (var server = new NamedPipeServerStream(pipeName))
-            using (var client = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation))
+            using (var client = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.None, tokenImpersonationLevel))
             {
                 Task serverTask = server.WaitForConnectionAsync();
 
                 client.Connect();
                 await serverTask;
 
                 bool ran = false;
-                server.RunAsClient(() => ran = true);
-                Assert.True(ran, "Expected delegate to have been invoked");
+                if (tokenImpersonationLevel == TokenImpersonationLevel.None)
```

Comment:
Just curious, why does `.None` cause an exception? And does this hit both the two exceptions that you fixed above?

Reply:
`None` causes the exception because there is no impersonation so the windows API, `ImpersonateNamedPipeClient`, fails. 

No, it only hits the path in which it fails to impersonate. The other branch is hit when it fails to revert to "self" (I don't know how to force that to fail).